### ISSUE TITLE
No exported member 'default'

### ISSUE THREAD

**Issue by sarbogast** [Quintor/angular-truffle-box#4]
When I try to run `ng serve`, I get the following errors:

```
ERROR in /Users/sarbogast/dev/DVoxApp/src/app/util/web3.service.ts (2,9): Module '"/Users/sarbogast/dev/DVoxApp/node_modules/web3/index"' has no exported member 'default'.
ERROR in /Users/sarbogast/dev/DVoxApp/src/app/util/web3.service.ts (2,31): Module '"/Users/sarbogast/dev/DVoxApp/node_modules/web3/index"' resolves to a non-module entity and cannot be imported using this construct.
```

I'm using the following versions:
- truffle 3.4.11
- solidity 0.4.15
- angular/cli: 1.4.9
- typescript: 2.5.3

**Comment by pimotte** [comment]
I'm not sure if it's necessary, but can you also check the version of web3 in package-lock.json (the direct dependency, not the dependency of one of the others)

**Comment by sarbogast** [comment]
```
"web3": {
      "version": "1.0.0-beta.24",
      "resolved": "https://registry.npmjs.org/web3/-/web3-1.0.0-beta.24.tgz",
      "integrity": "sha1-C2U9alD0B7N/WTNx+27hvTofJ8I=",
      "dev": true,
      "requires": {
        "web3-bzz": "1.0.0-beta.22",
        "web3-core": "1.0.0-beta.24",
        "web3-eth": "1.0.0-beta.24",
        "web3-eth-personal": "1.0.0-beta.24",
        "web3-net": "1.0.0-beta.24",
        "web3-shh": "1.0.0-beta.24",
        "web3-utils": "1.0.0-beta.24"
      }
    },
```

**Comment by pimotte** [comment]
I've pushed a fix at https://github.com/Quintor/angular-truffle-box/tree/bugfix/remove-web3-dep

Can you check out that branch and see if it works? (I can't imagine it doesn't, as this removes the entire web3 dependency)

**Comment by sarbogast** [comment]
So if you remove this dependency, does it mean it uses the same web3 as the one depended on by truffle-contract and so on?

**Comment by pimotte** [comment]
Ehm, it uses the provider that's tagged on to the version embedded in MetaMask, and I suspect truffle wraps that provider with it's own version of web3, but I haven't checked that

**Comment by sarbogast** [comment]
I mean in web3.service.ts

**Comment by sarbogast** [comment]
Anyway, now the compiler seems happy, but I have another weird error. Instead of checking out your branch, I simply removed the dependency in package.json, deleted package-lock.json and node_modules and re-installed depedencies. But now when I try to compile my contracts, I have another weird error:

```
Error: Cannot find module 'require-from-string'
    at Function.Module._resolveFilename (module.js:470:15)
    at Function.Module._load (module.js:418:25)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/usr/local/lib/node_modules/truffle/node_modules/solc/wrapper.js:2:25)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
```

Does that ring a bell to you?

**Comment by pimotte** [comment]
Yeah, so previously I would instantiate a new Web3 instance with the provider on the page, then use that instance to get the provider and use the .eth.getAccounts function. Now, I skip the step of creating the new instance and just use the MetaMask-exposed version directly.

**Comment by sarbogast** [comment]
Will this still work if the page is loaded in something like Mist?

**Comment by pimotte** [comment]
I've never tried it, but I'm pretty sure Mist injects web3 in the same way, but I've never checked, because when I looked I couldn't find a working version of Mist

**Comment by sarbogast** [comment]
OK, so I'm still left with my truffle compile error above. I'll investigate, thanks.

**Comment by pimotte** [comment]
Since the truffle compile is from the globally installed truffle you might try reinstalling that.

**Comment by sarbogast** [comment]
Reinstalling truffle solved my problem, but I'm not super comfortable with your change that removed the dependency on web3 entirely, since a lot of your code in the Web3Service class is counting on the fact that it's there. For example, this part of the code assumes that we are connecting to web3 1.0, doesn't it?

```
private refreshAccounts() {
    this.web3.eth.getAccounts((err, accs) => {
      if (err != null) {
        alert('There was an error fetching your accounts.');
        return;
      }

      // Get the initial account balance so it can be displayed.
      if (accs.length === 0) {
        alert('Couldn\'t get any accounts! Make sure your Ethereum client is configured correctly.');
        return;
      }

      if (!this.accounts || this.accounts.length !== accs.length || this.accounts[0] !== accs[0]) {
        console.log('Observed new accounts');
        this.accountsObservable.next(accs);
        this.accounts = accs;
      }

      this.ready = true;
    });
  }
```

This asynchronous call to web3.eth.accounts looks very 1.0-ish to me.

**Comment by pimotte** [comment]
Mmm. You do have a point. So maybe we should be wrapping it after all? Or do you have another idea? The thing is that https://github.com/Quintor/angular-truffle-box/tree/bugfix/remove-web3-dep suggests that the current usage will be broken, but the way it used to be looked kind of sketchy as well.

**Comment by sarbogast** [comment]
No other idea at this point. I just think we should be able to control which version of web3 we are using, especially in these times of such great upcoming change in web3's API.

**Comment by pimotte** [comment]
I've reverted the changes and pinned web3.js to 1.0.0-beta.23 which doesn't seem to have the problem you've experienced. The style of import we're using was also causing some issues when I tried to test, so if you have thoughts on that, they are very welcome.

**Comment by sarbogast** [comment]
Did you push your changes? Because I can't see it right now. And I can confirm that the MetaCoin sample was not working at all.

**Comment by pimotte** [comment]
I have pushed my changes to master

**Comment by sarbogast** [comment]
Weird because the latest commit I see on this repository is this one, the one you made about 15 hours ago to remove the dependency on web3: https://github.com/Quintor/angular-truffle-box/commit/242eec68c0eabf19a834169fffe1e96c8f3348de

**Comment by pimotte** [comment]
My bad, you're right. I pushed now. This should fix the issues for now, but we should be looking into why the import is failing in 1.0.0-beta.24, in case that makes it into the definitive 1.0.0 release of web3. Or possibly use the truffle included web3 to abstract?

**Comment by sarbogast** [comment]
I think it would make sense to create an angular-web3 wrapper around web3 that would be independent from the angular-truffle-box project. We could include some abstractions like your Web3Service and so on. And it could be imported in the angular-truffle-box as a dependency, or used in any existing Angular project. What do you think?

**Comment by pimotte** [comment]
I think conceptually it's a good solution, but I'm not sure what the interface would be. Would it just be an service that provides a Web3? How would we hook the contract abstractions, which are application dependent?

**Comment by sarbogast** [comment]
Not sure either but I'll look into it as I'm working on my own sample application for the Devoxx conference where we would like to showcase your box. I'll create pull requests as I move forward.

**Comment by sarbogast** [comment]
I just created a first pull request simply to clean up the code of the box and remove all the warning that IntelliJ and tslint were giving me: https://github.com/Quintor/angular-truffle-box/pull/5

**Comment by pimotte** [comment]
Awesome:)

I've made some issues for stuff that has been on my backlog, but I haven't gotten to. Feel free to create issues of what you're working on so that there's no double work done

**Comment by pimotte** [comment]
Fixed with #11
